compiler/src/dotty/tools/dotc/parsing/Parsers.scala

```diff
@@ -553,19 +553,27 @@ object Parsers {
     def inDefScopeBraces[T](body: => T, rewriteWithColon: Boolean = false): T =
       inBracesOrIndented(body, rewriteWithColon)
 
-    /** part { `separator` part }
-     */
-    def tokenSeparated[T](separator: Int, part: () => T): List[T] = {
+    /** part { `,` part }
+      * @param expectedEnd If set to something other than [[EMPTY]],
+      *                    assume this comma separated list must be followed by this token.
+      *                    If the parser consumes a `part` that is not followed by a comma or this expected
+      *                    token, issue a syntax error and try to recover at the next safe point.
+      */
+    def commaSeparated[T](part: () => T, expectedEnd: Token = EMPTY): List[T] = {
       val ts = new ListBuffer[T] += part()
-      while (in.token == separator) {
+      while (in.token == COMMA) {
         in.nextToken()
         ts += part()
       }
+      if (expectedEnd != EMPTY && in.token != expectedEnd) {
```

Comment:
The main change.

Reply:
So in case we have more that one unexpected token for example `foo(1 2 3, 4, 5)` this will not try to  parse the rest of the param clause?

Maybe it makes sense to try to find the comma or end of line?

Reply:
`syntaxErrorOrIncomplete` skips to the next safe pint as a side effect, so we will see `2` is not a `)`, issue an error, then jump to `,` and keep parsing. I'll add a third digit to one of the test cases to be sure.